nucliadb_node/src/shards/providers/shard_cache/resource_cache.rs

```diff
@@ -0,0 +1,427 @@
+    Wait(ResourceWaiter),
+}
+
+pub struct ResourceCache<K, V> {
+    live: LruCache<K, Arc<V>>,
+    eviction: HashMap<K, Weak<V>>,
+    capacity: Option<NonZeroUsize>,
+    loading: HashMap<K, Arc<Waiter>>,
+}
+
+impl<K, V> ResourceCache<K, V>
+where K: Eq + Hash + Clone + std::fmt::Debug
+{
+    #[allow(dead_code)]
+    pub fn new_with_capacity(capacity: NonZeroUsize) -> Self {
+        ResourceCache {
+            capacity: Some(capacity),
+            live: LruCache::unbounded(),
+            eviction: HashMap::new(),
+            loading: HashMap::new(),
+        }
+    }
+
+    pub fn new_unbounded() -> Self {
+        ResourceCache {
+            capacity: None,
+            live: LruCache::unbounded(),
+            eviction: HashMap::new(),
+            loading: HashMap::new(),
+        }
+    }
+
+    // Try to get an entry from the cache
+    // 1. If it's present, we return Cached(entry). Consumer can use it.
+    // 2. If it's not in the cache, we return Load(guard). Consumer should load it and call
+    //    cache.loaded with the guard.
+    // 3. If it's being loaded concurrently, we return Wait(waiter). Consumer should wait using the
+    //    waiter and then retry the get.
+    pub fn get(&mut self, id: &K) -> CacheResult<K, V> {
+        if let Some(v) = self.get_cached(id) {
+            return CacheResult::Cached(v);
+        }
+
+        if let Some(wait) = self.loading.get(id) {
+            if wait.finished() {
+                // The loading process finished but is still in the loading tree
+                // This happens when the guard was dropped without saving an object
+                // This is an error loading, we can retry by returning a load object
+                self.loading.remove(id);
+            } else {
+                return CacheResult::Wait(ResourceWaiter(Arc::clone(wait)));
+            }
+        }
+
+        let waiter = Arc::new(Waiter::default());
+        self.loading.insert(id.clone(), Arc::clone(&waiter));
+        CacheResult::Load(ResourceLoadGuard {
+            waiter,
+            key: id.clone(),
+        })
+    }
+
+    pub fn loaded(&mut self, guard: ResourceLoadGuard<K>, v: &Arc<V>) {
+        self.loading.remove(&guard.key);
+        self.insert(&guard.key, v);
+        drop(guard);
+    }
+
+    pub fn remove(&mut self, k: &K) {
+        self.live.pop(k);
+    }
+
+    fn get_cached(&mut self, id: &K) -> Option<Arc<V>> {
+        if let Some(v) = self.eviction.get(id).and_then(Weak::upgrade) {
+            self.insert(id, &v);
+            return Some(v);
+        }
+        self.live.get(id).cloned()
+    }
+
+    pub fn insert(&mut self, k: &K, v: &Arc<V>) {
+        if self.live.len() >= self.capacity.unwrap_or(NonZeroUsize::MAX).into() {
+            self.evict();
+        }
+        self.live.push(k.clone(), Arc::clone(v));
+    }
+
+    fn evict(&mut self) {
+        if let Some((evicted_k, evicted_v)) = self.live.pop_lru() {
+            println!("Evict {evicted_k:?}");
```

Comment:
I dont think we want this println in the actual code. Right?

Reply:
Uh, it shouldn't. I missed it. Thanks! Good :eye: